spring-web/src/main/java/org/springframework/http/server/reactive/TomcatServerHttpRequest.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.http.server.reactive;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import javax.servlet.AsyncContext;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.catalina.connector.CoyoteInputStream;
+import org.apache.catalina.connector.RequestFacade;
+import org.apache.coyote.Request;
+
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.core.io.buffer.DataBufferFactory;
+import org.springframework.core.io.buffer.DataBufferUtils;
+import org.springframework.http.HttpHeaders;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * Adapt {@link ServerHttpRequest} to the Tomcat {@link org.apache.coyote.Request}.
+ *
+ * @author Violeta Georgieva
+ * @author Brian Clozel
+ * @since 5.1
+ */
+final class TomcatServerHttpRequest extends ServletServerHttpRequest {
+
+	private static final Field COYOTE_REQUEST_FIELD = ReflectionUtils.findField(RequestFacade.class, "request");
+
+	private final int bufferSize;
+
+	private final DataBufferFactory factory;
+
+	static {
+		ReflectionUtils.makeAccessible(COYOTE_REQUEST_FIELD);
+	}
+
+	public TomcatServerHttpRequest(HttpServletRequest request, AsyncContext context,
+			String servletPath, DataBufferFactory factory, int bufferSize)
+			throws IOException, URISyntaxException {
+
+		super(createTomcatHttpHeaders(request), request, context, servletPath, factory, bufferSize);
+		this.factory = factory;
+		this.bufferSize = bufferSize;
+	}
+
+	private static HttpHeaders createTomcatHttpHeaders(HttpServletRequest request) {
+		Request tomcatRequest = ((org.apache.catalina.connector.Request) ReflectionUtils
+				.getField(COYOTE_REQUEST_FIELD, request)).getCoyoteRequest();
```

Comment:
Tomcat is the only server implementation that hides everything behind a facade; without reflection, I can't access the native Coyote request and its headers map.

Reply:
Can't we ask Tomcat to open their headers API for this use case, and offer a getter? I think that using reflection for this is taking things a bit too far, even if that means going back to the plain servlet approach.